Discussion: eBPF program & userspace program repo (#10)
l3af-project/l3af-arch: discussions/prog_repo.md

# L3AF eBPF Package Repository

The concept of a L3AF eBPF Package Repository is to create a location where
eBPF Programs from any trusted party can be uploaded and made available for others to download.

In the context of L3AF, we define an eBPF Package as a kernel space program with
an optional, cooperative user space program.

# What should we name it

"eBPF Package Repository" is the name that has been chosen by the Technical Steering Committee.

# Is the eBPF Package Repository part of the L3AF Project

Firstly, we define the "L3AF Project" as the L3AF open-source project that exists
within The Linux Foundation. When we say "L3AF", we are referring to the entire L3AF Project.
This is not to be confused with:

- The `l3af-project` GitHub organization
- The L3AFD daemon, which is a major component of the L3AF Project.

We consider L3AF to be an entire ecosystem that aims to provide eBPF Programs as a service.
We've otherwise phrased this as "complete lifecycle management of eBPF programs".
This definition includes:

- The L3AFD daemon, an orchestrator that provides APIs to launch and manage eBPF programs on a node

> - The L3AF eBPF Package Repository

Reviewer: Per discussion on line 119, maybe add the word "Public" here, since the L3AFD daemon can work with other package repositories too.   Indeed, if there is any actual "code" or scripts for such a repository, then we should separate implementation from policy.  I.e., implementation would be a repo that could be used by anyone when creating their own package repository, and policy is anything specifically for the L3AF Public eBPF Package Repository that may not apply to local repositories (e.g., the fact that you're proposing having to contribute source code instead of byte code is in my opinion, something that varies by repository).

Reviewer: Here’s my issues with requiring source code for all packages in the L3AF repo:

1)	If the authoritative version of ebpf source is already in some other existing repo, such as it is for Cilium, Calico, etc. today, then adding a copy under L3AF has both efficiency issues (why copy?  Which one is authoritative?  How do you submit a patch?), as well as possible governance issues (is the licensing model compatible?  Is the CLA/DCO process compatible?  Is the security vulnerability reporting process compatible?  Etc.) since it’s in two different open source governance organizations.
2)	If the tooling requires source code, rather than byte code, then there can be cases where the tooling won’t work for use in enterprise-private package repos if enterprises want to use their repo to contain both OSS packages and commercial packages they may not have source code for, so that they can use common L3AF tooling and program lifecycle management for all their ebpf needs.  This would result in multiple package repo implementations instead of reuse.

Using byte code in the repository, with recommending a pointer to where source can be found, would avoid both types of issue above.  This is why I think the initial version should do that instead of source.

Reviewer 2: My comment was really around the initial version and the limited set of (seed) programs added to this repo.  I don't think it is feasible to review / accept eBPF packages by l3af maintainers without the source code. 

Once there are more checks and gates in place, I don't think we should require source code and byte code alone would be sufficient.

Reviewer: > I don't think it is feasible to review / accept eBPF packages by l3af maintainers without the source code.

My opinion is that manual review is not scalable.   If the "public" (or whatever) repo is limited to only what can be done via manual review, then I would expect it will always only contain a small number of small programs, and additions will take a long time to get reviewed, and the reviewers will have a long backlog of requests, and people will get displeased with the length of time it takes to get anywhere, and go elsewhere to a different "public" repo that is automated and does not attempt to make promises about security vetting, but leaves that responsibility to those pulling from the repo, and reinforce the point that one should never point l3afd at the "public" repo only at one's own vetted repo (which would contain a small number of programs by comparison).  Personally I would rather have the initial version do the latter rather than doing something that requires much more effort and still doesn't solve the security concerns.

Reviewer 2: I agree that manual review is not scalable and not something we should offer to the public until we have the automation in place. What if we keep the name "L3AF eBPF Package Repository" repository for now and then change it to "Public L3AF eBPF Package Repository" when we are ready with the automation?  I think this is a good initial place to store the build artifacts (byte code) for several of the eBPF programs contributed already under L3AF (xdp-root, ratelimiting, connection-limit). In my opinion it seems cleaner to store these here vs. in the l3afd repository.  When we are ready to open this to the public they will already be added and can be reference examples for other folks to contribute their own eBPF packages to this repo.

Reviewer 3: I think it is important to distinguish between the initial and future version spec of the "eBPF Package Repository". 

In the initial version spec, the proposal is to simply have a GitHub repository to store eBPF program source code. This is to make sure that all the contributed eBPF programs can be made available in a single repository. The proposal acknowledges that the initial version won't have automation in place, and so all submissions will be manually reviewed and published in this repo. The review process will also ensure that the code submissions conform to L3AF's eBPF program chaining mechanics. In this context, it may not be possible to review and accept eBPF packages without source code.

The future version builds on top of the initial version and aims to include a fully automated build-release system that can perform reviews and build/upload the package to a storage repo. The proposal for future version discusses the option of allowing contributors to upload the eBPF package without sharing the source code. It also aims to address various other challenges such as security (vulnerability detection, signing, semantic code analysis) and portability. The eBPF Package Repository Committee will discuss the future spec in detail to draw agreements and can choose to add the word "public" to "eBPF Package Repository".  

Hope this helps clarify!

Reviewer 4: Opened issue for further discussion here https://github.com/l3af-project/eBPF-Package-Repository/issues/2

- Programs within the eBPF Package Repository

Note, however, that we have no intention of limiting the creation of other public or private
eBPF Package repositories.

We recognize that it may make sense to migrate the eBPF Package Repository

> (eBPF Programs within the repository) out of the L3AF Project and into its own

Reviewer: ```suggestion
(and any eBPF Packages within the repository) out of the L3AF Project and into its own
```
Not sure what was meant by the parenthetical phrase, is this what was meant?

Linux Foundation project in the future. Doing this initially may not make sense (due to L3AF-specific
eBPF program chaining implementation for networking programs), but as L3AF and other projects mature,
a platform-agnostic repository could be useful for multiple projects.

# What should an initial version look like

> This section examines simple ways to create a location where eBPF Programs can be uploaded

Reviewer: ```suggestion
This section examines simple ways to create a location where eBPF Packages can be uploaded
```

Reviewer 4: @dthaler In the initial version we will be hosting eBPF program source code in a github repo and not the eBPF Packages.

and made available for others to download.

## A GitHub repository may be sufficient

We would like to leverage GitHub:

- A new GitHub repository will be created to store eBPF program source code
- All submissions will be manually reviewed by the L3AF team
- Once approved, programs will be published in the eBPF Package Repository

Another important thing to note is that, initially, code submissions will need
to conform to L3AF's eBPF program chaining mechanics.

## To Build or Not to Build

Because of the overhead and support requirements of a full build system, it may not be
feasible to build the eBPF program source code in the initial version. However, we believe

> this would greatly hinder adoption from both contributors and users. Our proposal, therefore,

Reviewer: what is "this" referring to?   having to build from source?  The word "However" seems confusing if so.

Reviewer 4: @dthaler "this" means "not having to build from source".

Reviewer: ```suggestion
not being able to build from source would greatly hinder adoption from both contributors and users. Our proposal, therefore,
```

is to have the repository’s initial version include scripts (e.g. Dockerfile for build system images),
and steps to build eBPF Programs locally.

## eBPF Package Repository of the future

Future versions could build on top of the foundation laid by the initial version.

# What should a future version look like

> This section examines simple ways to create a location where eBPF Programs from trusted parties

Reviewer: ```suggestion
This section examines simple ways to create a location where eBPF Packages from trusted parties
```

can be uploaded and made available for others to download. This can also enable users to provide
rating, reviews, and tags to the packages.

### Build Process

> Contributed eBPF programs can be built on common images for Linux and other platforms.

Reviewer: ```suggestion
Contributed eBPF Packages can be built on common images for Linux and other platforms.
```

> Contributors are expected to build their code using the most recent image used by our build system.

Reviewer: ```suggestion
Contributors are expected to build their code using the most recent images used by our build system.
```

Contributors could have a choice to build on Linux only or other platforms.

Build artifacts (i.e., eBPF program bytecode and user space binaries) could be stored

> in a public file storage repository. Users could download the artifacts directly

Reviewer: Aren't those inside the eBPF Packages uploaded?

Reviewer 4: @dthaler Yes, here artifacts mean nothing but eBPF packages.

from this repository. In fact, such a repository could also be considered the
eBPF Package Repository, and the source repository could be a separate entity.

### Portability

It is hard to discuss a build system without broaching the topic of portability.
The eBPF portability story has improved recently with the introduction of eBPF CO-RE on Linux.
Previous issues with portability and the eBPF CO-RE solution are explained here:

https://nakryiko.com/posts/bpf-portability-and-co-re/

> Because an eBPF Package Repository is a place from where contributed eBPF programs (byte code)

Reviewer: As far as we currently know, byte code won't be sufficient for cases like hypervisor-enforced code integrity.  For such cases, the package would need to contain native code not ebpf byte code.

Reviewer 4: @dthaler We can explore this use case going forward.

Reviewer 2: Perhaps we can remove the (byte code) designation here and say while compiled is the target, we can leave the door open to interpreted (native) code in the eBPF package repo

can be downloaded to run on a variety of kernel versions, we propose that the repository follows
best practices for compatibility, such as using eBPF CO-RE for Linux. Similar best practices
can be followed on non-Linux platforms as they mature and become available.
The user space components of eBPF programs pose a separate, complicated portability obstacle,
which is compounded by the desire of the L3AF project to support user space programs in multiple

> languages. For the initial version of the repository, it should be sufficient to build

Reviewer: This line contradicts line 61 which to me implies the uploader compiles locally before submitting to the repository, which I agree with.

Reviewer 4: @dthaler This section talks about the future version of the eBPF Package Repository, we will have packages built for x86_64 platforms only and decide about other platforms in future on need basis.

Reviewer: If I understand your comment, then you'd want to reword line 61 since right now the document appears to contradict itself.

(of a compiled language) and unit test for the x86_64 platform. The user space component
should document any installation dependencies it has (e.g., MySQL, Grafana, Python libraries, etc.).
Contributors to the repository would be responsible to provide the necessary build scripts
and configuration.

### Multiple Versions

Regardless of whether the package is built from source or not, eBPF programs should use semantic
versioning. The repository should then host previous versions (in addition to the current version)
in some reasonable manner.

### Automated Reviews

There can be a mechanism to review source code, using tools to detect like code formatting errors,
complexity, code issues, and code duplication. Similarly, vulnerability detection
scanners can be used for artifacts before uploading.

## Alternative to Hosting and Building Source Code

An alternative to the L3AF Project hosting and building contributed source code would be for

> contributors to submit or self-host a signed package containing their eBPF programs and any

Reviewer: Doesn't line 61 say this alternative is what would be the initial version?  In which case it's not an "alternative" per se?  Or am I misunderstanding the earlier section?

Reviewer 4: @dthaler Initial version of eBPF Package Repository is to host eBPF Program source code in a github repo. But this section describes if someone does not want to share the source code, and still wants to share the eBPF Package, then it should be supported.

Reviewer: In that case you'd probably want a more specific name for the particular eBPF Package Repository (e.g., "L3AF Public Package Repository") with that policy, since that may just discourage some from using that repository and instead use their own.  The document already allows for that, but I would prefer to separate the concept of package repositories in general, from the policies around a very specific one that may or may not be used, up to the admin of any particular l3afd installation.

Reviewer 4: This is just an alternative hosting approach, we can decide on this in future on need basis.

Reviewer 4: This issue will be addressed here https://github.com/l3af-project/eBPF-Package-Repository/issues/2

documentation. The L3AF daemon, by default, would only download and run eBPF programs signed
by trusted parties. This provides a nice segue for us to go into the signing section.

## Signing and Verification

There are three different layers of signing that should be discussed, the signing of the
kernel eBPF program (i.e., byte code), the signing of the eBPF programs package (i.e., an archive
containing the built eBPF program and any associated user space programs) and
signing for Hypervisors (i.e., Virtualization-based Security).

### Signing and verifying eBPF programs

There are efforts underway in the eBPF community for how to sign and verify
Linux eBPF programs, this enables JIT’ed native code:

https://lore.kernel.org/bpf/20211203191844.69709-1-mcroce@linux.microsoft.com/
https://linuxplumbersconf.org/event/11/contributions/947/

Similarly, signing mechanism for other platforms are also being explored.

L3AF can use the best practices that are established for signing and verifying eBPF programs.

### Signing and verifying the eBPF Program package

> At a higher level, we also plan to sign and verify eBPF Package.

Reviewer: ```suggestion
At a higher level, we also plan to sign and verify eBPF Packages.
```

The package can be signed by its creator using eBPF Package Repository generated/provided
trusted keys. L3AF can, by default, verify the package comes from a trusted source before executing

> any programs in the package

Reviewer 2: I think we would want to only allow l3af project reviewed and accepted programs to get signed with the eBPF repository trusted key. I would suggest l3af by default does not trust any 3rd party / trusted party signed packages and it would be required to configure l3afd to trust packages from a certain party.  For example, if Microsoft has signed eBPF packages, you should have to configure l3afd to explicitly allow signed packages from Microsoft.

Reviewer: Agree, and to allow packages from the L3AF Public Repository (or whatever it's called) you should have to configure l3afd to explicitly allow signed packages from that repository.

### Signing for Hypervisors

This approach needs to be explored based on hypervisor operating systems security policies.

### Define trust

> In this context, trusted user and contributor is defined as authorised registered user to the
> system, who has credentials to login with RBAC restrictions. Contributor can be given access

Reviewer: ```suggestion
In this context, a trusted user and contributor is defined as an authorised user registered to the
system, who has credentials to log in with RBAC restrictions. A contributor can be given access
```
grammar

> # L3AF Kernel Function Marketplace

Reviewer 5: In the 2022-01-05 call we discussed renaming this a `package manager`.

The concept of a L3AF Kernel Function Marketplace is to create a location where
Kernel Functions from any trusted party can be uploaded and made available for
others to download.

> In the context of L3AF, we define a Kernel Function a kernel eBPF program with

Author: The name "Kernel Function" for this concept is probably a misnomer. Please discuss alternative names here.

Reviewer 5: We'd like this name to (a) not include "kernel", (b) not be limited to "functions", (c) be quick, easy, descriptive, (d) not reinvent the wheel for the sake of being "cute" or "clever". Some ideas off the top of my head…

* App: commonly used term; does it imply a self-contained thing or would it work as well for chainables?
* Package: works with `package manager`; implies it's a thing that contains other things, which these will be? I think?

However, this brings up a question: When you look at one of these things that the package manager, uh, manages, what do you see? Is it a collection of related files or a definition/manifest that can build the collection or…? Perhaps we should get a shared/common idea of _what's_ being distributed before we go much further on _how_ they get distributed?

Author: `eBPF Package` seems like a reasonable name imo

Reviewer: "ebpf package" is fine with me for the thing that can include both app code and ebpf program code.

Reviewer 3: I agree,  we can probably look to replace "Kernel Function" with "eBPF Program" and we can also call this "repository".. which would make this eBPF Program Repository.

Reviewer 3: Catching up on the above comments - - We can call it `eBPF package`; the reason I preferred the name `eBPF program` is that this repo may contain many simple eBPF programs that leverage kprobes, uprobes, and tracepoints. And so, an eBPF program will really be the LCD here.

an optional, cooperative userspace program.

There are many things to consider. At the highest level, the purpose of this
discussion is to arrive an conclusions for:

- What should we name it
- Is the Kernel Function Marketplace part of the L3AF project?
- What should an initial version look like
- What should a more mature version look like

# What should we name it

> "L3AF Kernel Function Marketplace" has been the name used up to this point and

Author: The name "Kernel Function Marketplace" seems generally disliked. Please discuss alternatives here.

Reviewer 5: Per above (which I probably should delete in preference to this comment but meh), in the 2022-01-05 call the general consensus was that `package manager` was a better name than `marketplace`. It certainly has my vote.

Author: When I hear the term `package manager` I'm thinking of things like APT, yum, zypper, etc.  I.e., I think a `package manager` is generally thought of a tool for installing, update, and removing packages.

I think I'd prefer the term `eBPF Package Repository`.

Then, perhaps L3AFD could be considered an eBPF package manager.

Reviewer: Agree that "manager" is not the right word for the repository, it's the word for an agent that runs on the node and can use a repository.  +1 to "eBPF Package Repository".

not make sense (due to L3AF-specific eBPF program chaining mechanics, for
example), but as L3AF and other projects mature, a platform-agnostic
marketplace could be useful for multiple projects.

# What should an initial version look like

This sections examines relatively simply ways that we can create a location
where Kernel Functions from trusted parties can be uploaded and made available
for others to download.

## A GitHub repository may be sufficient
 
We could leverage GitHub:
- A new GitHub repository could be created to store kernel function source code
- All submissions could be reviewed by the L3AF team
- Once approved, users could download the kernel function (source and any build
  artifacts)

If we use this approach, this GitHub repo could be created under
github.com/leaf-project.

Kernel functions could be arranged by file path and categorized in any number
of ways. For example, using a schema such as: 

`/{Program Category}/{Program Subcategory}/{Submitter}/{Program Name}`

Could translate to a kernel function being stored at:

`/Security/limits/Walmart/ratelimit`

Another important thing to note is that, initially, code submissions will need

> to conform to L3AF's eBPF program chaining mechanics.

Reviewer 2: I think it probably makes sense to investigate different chaining semantics based on underlying platform support / requirements in the future

https://github.com/libbpf/libbpf#bpf-co-re-compile-once--run-everywhere

Userspace programs would also ideally build on a common image.

Build artifacts (i.e., eBPF program bytecode and userspace binaries) would be
stored in a public file storage repository. Users could download the build
artifacts directory from this repository. In fact, such a repository could also
be considered the Kernel Function Marketplace, and the source repository could
be a separate entity.

After building from source, we would then store the build artifacts into an
package file and sign it. The file would be signed by the L3AF Project.

### Multiple Versions

Regardless of whether the L3AF project builds programs from source or not,
Kernel Functions should use semantic versioning. The marketplace should then
host previous versions (in addition to the current version) in some reasonable
manner.

## Alternative to Hosting and Building Source Code

An alternative to the L3AF project hosting and building contributed source code
would be for contributors to submit or self-host a signed package containing
their Kernel Function and any documentation. The L3AF daemon, by default, would

> only download and run Kernel Functions signed by trusted parties.

Reviewer: Elaborate: how would L3AFD know who the acceptable trusted parties are?  Does it have a set of public keys?  Where would it store them?  How would you update it to add/remove a trusted party? How would you do key rollover?

## Kernel Function Signing and Verification

> There are two different layers of signing that should be discussed: the signing

Reviewer: There's actually 3 layers in general.  I.e., there's a difference between signing eBPF byte code, and signing JIT'ed native code.   The former might be used to bypass the Verifier step if verification is a pre-requisite for signing.   The latter would be required to work with Type 1 hypervisor-enforced kernel code integrity (e.g., as supported in Hyper-V).

of the kernel eBPF program and the signing of the Kernel Function package (i.e.,
an archive containing the built eBPF program and any associated userspace
programs).

### Signing and Verifying eBPF programs

There are efforts underway in the eBPF community for how to sign and verify
eBPF programs:

https://lore.kernel.org/bpf/20211203191844.69709-1-mcroce@linux.microsoft.com/
https://linuxplumbersconf.org/event/11/contributions/947/

L3AF will use the best practices that are established for signing and verifying
eBPF programs.

### Signing and Verifying the Kernel Function package

At a higher level, we also plan to sign and verify Kernel Function packages.

> The package will be signed by its creator using a Private Enterprise Number

Reviewer: You sign things with a key.  A PEN is just a number and can't sign things.
